### PR TITLE
Allow creators to disable hand touch from javascript

### PR DESCRIPTION
**Pull request by luiscuenca**
This PR add the following API calls so creators can enable/disable the hand touch effect globally or by entityID:
```
MyAvatar.requestDisableHandTouch()
MyAvatar.requestEnableHandTouch()
MyAvatar.disableHandTouchForID(entityID)
MyAvatar.enableHandTouchForID(entityID)
```

https://highfidelity.manuscript.com/f/cases/16728/Finger-movement-script-interferes-with-content-piano

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11353/HighFidelity-Beta-PR13619-84dac94.apk

**Comment by tapalisa** [comment]
If a content creator forgets to re-enable it, what happens? is the script permanently turned off for the user?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11353/HighFidelity-Beta-PR13619-84dac94.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11353/HighFidelity-Beta-Interface-PR13619-84dac94.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11353/HighFidelity-Beta-PR13619-84dac94.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11353/HighFidelity-Beta-Interface-PR13619-84dac94.exe

**Comment by luiscuenca** [comment]
good point @tapalisa 
I added an option that disable hand touch by entity ID. I think that method will be perfect for what we are looking for. 
I kept the possibility of enable/disable hand touch in case is necessary for certain avatars or situations where other scripts interfere with it.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11358/HighFidelity-Beta-PR13619-82a3b73.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11358/HighFidelity-Beta-PR13619-82a3b73.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11358/HighFidelity-Beta-PR13619-82a3b73.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11358/HighFidelity-Beta-Interface-PR13619-82a3b73.exe

**Comment by luiscuenca** [comment]
## TEST PLAN
- Wear an avatar with proper fingers. (i.e. Mathew)
- Create a **Box** and a **Sphere** and make sure you can touch them. (your fingers adapt to the surface)
- Open the console and write `MyAvatar.requestDisableHandTouch()`. Make sure that the touching effect is unable for both entities.
- `MyAvatar.requestEnableHandTouch()` should enable touching for all entities.
- Copy the **Sphere** ID _**sphereID**_ .(Create -> Select the **Sphere** -> ID field)
- Write in the console `MyAvatar.disableHandTouchForID("`_**sphereID**_`")` where _**sphereID**_ is the value you copied before. This should disable hand touch for the **Sphere** only. Make sure you can still touch the **Box**.
- Disable hand touch for the **Box**, with `MyAvatar.disableHandTouchForID("`_**boxID**_`")` . Get the boxID following previous steps. Make sure you can't touch any of them.
- Enable hand touch for the **Box** `MyAvatar.enableHandTouchForID("`_**boxID**_`")`. Make sure you can touch the **Box** but not the **Sphere**.
- Enable hand touch for the **Sphere** `MyAvatar.enableHandTouchForID("`_**sphereID**_`")`. Make sure you can touch both, the **Box** and the **Sphere**.

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Fail @Qliemillar 
Using MyAvatar.disableHandTouchForID or MyAvatar.enableHandTouchForID did not enable/disable the touch effect for entities. Both the Box and Sphere did not change properties when using the appropriate MyAvatar function with their ID's.
[hifi-log.txt](https://github.com/highfidelity/hifi/files/2202886/hifi-log.txt)

File: interface/src/avatar/MyAvatar.h [hunk 1]
```diff
@@ -505,6 +505,24 @@ class MyAvatar : public Avatar {
...
+    * @function MyAvatar.requestEnableHandTouch
+    */
+    Q_INVOKABLE void requestEnableHandTouch();
+    /**jsdoc
+    * @function MyAvatar.requestDisableHandTouch
+    */
+    Q_INVOKABLE void requestDisableHandTouch();
+    /**jsdoc
+    * @function MyAvatar.disableHandTouchForID
+    * @param {string} entityId
```
Comment (ctrlaltdavid) [inline interface/src/avatar/MyAvatar.h]: This function doesn't have an entityId parameters.

File: interface/src/avatar/MyAvatar.h [hunk 2]
```diff
@@ -505,6 +505,24 @@ class MyAvatar : public Avatar {
...
+    */
+    Q_INVOKABLE void disableHandTouchForID(const QString& entityId);
```
Comment (ctrlaltdavid) [inline interface/src/avatar/MyAvatar.h]: Parameter should be `const QUuid& entityID`. I.e., QUuid type and upper case "ID".

File: interface/src/avatar/MyAvatar.h [hunk 3]
```diff
@@ -505,6 +505,24 @@ class MyAvatar : public Avatar {
      * @returns {boolean} 
      */
     Q_INVOKABLE bool getHMDLeanRecenterEnabled() const { return _hmdLeanRecenterEnabled; }
+    /**jsdoc
+    * @function MyAvatar.requestEnableHandTouch
```
Comment (ctrlaltdavid) [inline interface/src/avatar/MyAvatar.h]: Descriptions of functions and parameters should be provided.

File: interface/src/avatar/MyAvatar.h [hunk 4]
```diff
@@ -1667,6 +1700,7 @@ private slots:
     bool _shouldLoadScripts { false };
 
     bool _haveReceivedHeightLimitsFromDomain { false };
+    int _disableHandTouchCount { 0 };
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.h]: This needs to be std::atomic<int> in order for ++ to be thread safe.

File: interface/src/avatar/MyAvatar.cpp [hunk 5]
```diff
@@ -262,6 +262,24 @@ void MyAvatar::setDominantHand(const QString& hand) {
     }
 }
 
+void MyAvatar::requestDisableHandTouch() {
+    _disableHandTouchCount++;
+    emit shouldDisableHandTouchChanged(_disableHandTouchCount > 0);
+}
+
+void MyAvatar::requestEnableHandTouch() {
+    _disableHandTouchCount = std::max(_disableHandTouchCount - 1, 0);
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: this is not thread safe

File: interface/src/avatar/MyAvatar.cpp [hunk 6]
```diff
@@ -262,6 +262,24 @@ void MyAvatar::setDominantHand(const QString& hand) {
     }
 }
 
+void MyAvatar::requestDisableHandTouch() {
+    _disableHandTouchCount++;
+    emit shouldDisableHandTouchChanged(_disableHandTouchCount > 0);
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: ++ is not thread safe unless _disableHandTouchCount is changed from an int to a std::atomic<int>

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Pass @qliacoombs @Qliemillar 
Touch can be enabled and disabled globally and on certain entities.

**Pull request merged**